src/Submission/ProjectSubmission.jsx

```diff
@@ -9,55 +9,105 @@ import './ProjectSubmission.less';
 import { useArboristUI } from '../configs';
 import { userHasMethodOnProject, isRootUrl, isProgramUrl, userHasSheepdogProgramAdmin, userHasSheepdogProjectAdmin } from '../authMappingUtils';
 
-const ProjectSubmission = (props) => {
-  // hack to detect if dictionary data is available, and to trigger fetch if not
-  if (!props.dataIsReady) {
-    props.onGetCounts(props.typeList, props.project, props.dictionary);
+class ProjectSubmission extends React.Component {
+  componentDidMount() {
+    this.props.fetchPrograms();
   }
 
-  // Passing children in as props allows us to swap in different containers -
-  // dumb, redux, ...
-  const MySubmitForm = props.submitForm;
-  const MySubmitTSV = props.submitTSV;
-  const MyDataModelGraph = props.dataModelGraph;
-  const displayData = () => {
-    if (!props.dataIsReady) {
-      if (props.project !== '_root') {
-        return <Spinner />;
-      }
-      return null;
+  componentDidUpdate() {
+    if (this.props.programList && !this.shouldDisplayProjectUIComponents(
+      this.props.project,
+      this.props.projectList,
+      this.props.programList,
+      this.props.userAuthMapping,
+    )) {
+      this.props.history.push('/not-found');
     }
-    return <MyDataModelGraph project={props.project} />;
-  };
-  const displaySubmissionUIComponents = (project, userAuthMapping) => {
-    if (
-      !useArboristUI
-      || (isRootUrl(project) && userHasSheepdogProgramAdmin(userAuthMapping))
-      || (isProgramUrl(project) && userHasSheepdogProjectAdmin(userAuthMapping))
-      || userHasMethodOnProject('create', project, userAuthMapping)
-      || userHasMethodOnProject('update', project, userAuthMapping)
-    ) {
-      return (
-        <React.Fragment>
-          <MySubmitForm />
-          <MySubmitTSV project={project} />
-        </React.Fragment>
-      );
+  }
+
+  shouldDisplayProjectUIComponents = (
+    project,
+    projectList,
+    programList,
+    userAuthMapping,
+  ) => {
+    if (useArboristUI) {
+      // check against arborist policies
+      return ((isRootUrl(project) && userHasSheepdogProgramAdmin(userAuthMapping))
+        || (isProgramUrl(project)
+        && programList.includes(project)
+        && userHasSheepdogProjectAdmin(userAuthMapping))
+        || Object.values(projectList).includes(project));
```

Comment:
Making sure I understand, when we get `projectList` from Peregrine, it only includes projects that a user has access to, right? So a user can only view the `commons.url/program-project` page if they have access to the program and project. (Or are a sheepdog project/program admin)

Reply:
Yes! I believe the `projectList` only contains the projects that user has access to. But seems like the `programList` would always contains all the program names in the sheepdog DB. I think this is the current peregrine behavior?

But if we turned the arborist UI on then if a user is not project admin then he/she cannot see any program pages (they are meaningless to them anyways i think)

Reply:
Update, I think I get it now -- I'm a user, I can view the `commons.url/program-project` page as long as program/project exists, and if I'm a sheepdog project admin, I can view `commons.url/program-newproject` even if `program/newproject` doesn't exist yet?

Reply:
actually i should say this:
- If arborist UI is ON, user will NOT be redirect to `/not-found` if 
1. the URL is `_root` and user has sheepdog PROGRAM admin (in this case user can see _root submission page)
2. the URL is `commons.url/program` and the program exists in sheepdog DB and user has sheepdog PROJECT admin (in this case user can see program submission page)
3. the URL is `commons.url/program-project` and the project exists in sheepdog DB and user has access to that project (in this case user can see project submission page)

Reply:
Note the above discussion only applies to whether a user will be redirected to `/not-found`. It does NOT include whether user can see submission UI components (the components inside red box)
![Screen Shot 2020-06-15 at 4 26 54 PM](https://user-images.githubusercontent.com/2475897/85064106-b3402380-b170-11ea-8855-224078f0a7ee.png)

The controlling logic of that UI component is at https://github.com/uc-cdis/data-portal/blob/580922bfed7b84ff867b08836155ba82d3141a74/src/Submission/ProjectSubmission.jsx#L78

Reply:
Gotchu, that all makes sense -- thanks for the explanation!